From df6cf403c8d33548ffcea6577a9bbf1845aa9cae Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH 281/296] Avoid strict mode warning when persisting sync state

We need to be sure that the write of the SyncManager state is complete
when broadcast of ACTION_SHUTDOWN is complete, so doing it
asynchronously is not an alternative.

This will remove the strict mode warning.
---
 .../com/android/server/content/SyncManager.java    | 23 +++++++++++++---------
 1 file changed, 14 insertions(+), 9 deletions(-)

diff --git a/services/core/java/com/android/server/content/SyncManager.java b/services/core/java/com/android/server/content/SyncManager.java
index 581e8c748e6..6839a90f9da 100644
--- a/services/core/java/com/android/server/content/SyncManager.java
+++ b/services/core/java/com/android/server/content/SyncManager.java
@@ -70,6 +70,7 @@ import android.os.PowerManager;
 import android.os.RemoteCallback;
 import android.os.RemoteException;
 import android.os.ServiceManager;
+import android.os.StrictMode;
 import android.os.SystemClock;
 import android.os.SystemProperties;
 import android.os.UserHandle;
@@ -414,15 +415,19 @@ public class SyncManager {
 
     private BroadcastReceiver mShutdownIntentReceiver =
             new BroadcastReceiver() {
-                @Override
-                public void onReceive(Context context, Intent intent) {
-                    Log.w(TAG, "Writing sync state before shutdown...");
-                    getSyncStorageEngine().writeAllState();
-
-                    mLogger.log(getJobStats());
-                    mLogger.log("Shutting down.");
-                }
-            };
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            final StrictMode.ThreadPolicy savedPolicy = StrictMode.allowThreadDiskWrites();
+            try {
+                Log.w(TAG, "Writing sync state before shutdown...");
+                getSyncStorageEngine().writeAllState();
+                mLogger.log(getJobStats());
+                mLogger.log("Shutting down.");
+            } finally {
+                StrictMode.setThreadPolicy(savedPolicy);
+            }
+        }
+    };
 
     private BroadcastReceiver mUserIntentReceiver = new BroadcastReceiver() {
         @Override
-- 
2.11.0

